### PR TITLE
Enable pre execution for skvbc_reply

### PR DESCRIPTION
**Pull request by situ-s**
Enable pre execution for skvbc_reply

### CODE REVIEW

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 1]
```diff
@@ -1580,9 +1582,14 @@ OperationResult PreProcessor::launchReqPreProcessing(const PreProcessRequestMsgS
...
+    resultLen = strlen(preProcessResultBuffer);
+    // Append the conflict detection block id and add its size to the resulting length.
+    memcpy(preProcessResultBuffer + resultLen, reinterpret_cast<char *>(&blockId), sizeof(uint64_t));
+    resultLen += sizeof(uint64_t);
+    LOG_DEBUG(logger(),
+              "Pre-execution operation successfully completed with errored result"
+                  << KVLOG(cid, reqSeqNum, clientId, reqOffsetInBatch, blockId, (uint32_t)preProcessResult));
     return preProcessResult;
   }
   if (request.outActualReplySize == 0) {
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.cpp:1595]: Here we need to fill the preProcessResultBuffer with en error "Empty execution data received" and fill the block at the end of the function, too.

File: bftengine/src/preprocessor/RequestProcessingState.cpp [hunk 2]
```diff
@@ -77,15 +77,6 @@ uint32_t RequestProcessingState::sizeOfPreProcessResultData() const {
...
 
-void RequestProcessingState::setupPreProcessResultData(OperationResult preProcessResult) {
-  memcpy(&primaryPreProcessResultData_, &preProcessResult, sizeof(preProcessResult));
-  primaryPreProcessResultData_ += sizeof(preProcessResult);
-  memcpy(&primaryPreProcessResultData_, &clientId_, sizeof(clientId_));
-  primaryPreProcessResultData_ += sizeof(clientId_);
-  memcpy(&primaryPreProcessResultData_, &reqSeqNum_, sizeof(reqSeqNum_));
-  primaryPreProcessResultLen_ = sizeOfPreProcessResultData();
-}
-
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/RequestProcessingState.cpp:88]: Why did you remove this code?

File: bftengine/src/preprocessor/RequestProcessingState.cpp [hunk 3]
```diff
@@ -96,16 +87,16 @@ void RequestProcessingState::handlePrimaryPreProcessed(const char *preProcessRes
...
     primaryPreProcessResultData_ = preProcessResultData;
     primaryPreProcessResultLen_ = preProcessResultLen;
   }
+
   primaryPreProcessResultHash_ = PreProcessResultHashCreator::create(
       primaryPreProcessResultData_, primaryPreProcessResultLen_, primaryPreProcessResult_, clientId_, reqSeqNum_);
 
   // In case the pre-processing failed on the primary replica, fill primaryPreProcessResultData_ by the result-related
   // information.
-  if (preProcessResult != OperationResult::SUCCESS) setupPreProcessResultData(preProcessResult);
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/RequestProcessingState.cpp:108]: Why did you remove this code?

**Comment by yuliasherman** [comment]
Please add description to the PR.

File: bftengine/src/preprocessor/PreProcessor.cpp (hunk 1 above)
Comment (situ-s) [reply bftengine/src/preprocessor/PreProcessor.cpp:1628]: done

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 4]
```diff
@@ -1411,7 +1411,8 @@ bool PreProcessor::registerRequestOnPrimaryReplica(const string &batchCid,
                                         clientReqMsg->requestSignature(),
                                         clientReqMsg->requestSignatureLength(),
                                         blockId,
-                                        clientReqMsg->spanContext<ClientPreProcessReqMsgUniquePtr::element_type>());
+                                        clientReqMsg->spanContext<ClientPreProcessReqMsgUniquePtr::element_type>(),
+                                        clientReqMsg->result());
```
[the remark this answers is not in the document]
Comment (situ-s) [reply bftengine/src/preprocessor/PreProcessor.cpp:1460]: its not for log. Its passed to PreProcessRequestMsg, otherwise by default will be unknown

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 5]
```diff
@@ -1580,9 +1582,14 @@ OperationResult PreProcessor::launchReqPreProcessing(const PreProcessRequestMsgS
   auto preProcessResult = static_cast<OperationResult>(request.outExecutionStatus);
   resultLen = request.outActualReplySize;
   if (preProcessResult != OperationResult::SUCCESS) {
-    LOG_ERROR(logger(),
-              "Pre-execution failed" << KVLOG(
-                  cid, clientId, reqOffsetInBatch, reqSeqNum, (uint32_t)preProcessResult, resultLen));
+    strcpy(preProcessResultBuffer, "Error in execution request");
```
[the remark this answers is not in the document]
Comment (situ-s) [reply bftengine/src/preprocessor/PreProcessor.cpp]: done

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 6]
```diff
@@ -1580,9 +1582,14 @@ OperationResult PreProcessor::launchReqPreProcessing(const PreProcessRequestMsgS
...
+    resultLen = strlen(preProcessResultBuffer);
+    // Append the conflict detection block id and add its size to the resulting length.
+    memcpy(preProcessResultBuffer + resultLen, reinterpret_cast<char *>(&blockId), sizeof(uint64_t));
```
[the remark this answers is not in the document]
Comment (situ-s) [reply bftengine/src/preprocessor/PreProcessor.cpp]: removed it

File: bftengine/src/preprocessor/RequestProcessingState.cpp (hunk 2 above)
Comment (situ-s) [reply bftengine/src/preprocessor/RequestProcessingState.cpp:88]: added back with change. Earlier & address was passed, and this caused issue in copying the data back to clientRequest.

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 7]
```diff
@@ -1623,13 +1624,10 @@ OperationResult PreProcessor::launchReqPreProcessing(const PreProcessRequestMsgS
...
-                  cid, clientId, reqOffsetInBatch, reqSeqNum, (uint32_t)preProcessResult, resultLen));
-    return preProcessResult;
-  }
+
   if (request.outActualReplySize == 0) {
+    strcpy(preProcessResultBuffer, "Executed data is empty");
+    resultLen = strlen(preProcessResultBuffer);
     preProcessResult = OperationResult::EXEC_DATA_EMPTY;
     LOG_ERROR(logger(),
               "Pre-execution failed" << KVLOG(cid, clientId, reqOffsetInBatch, reqSeqNum, (uint32_t)preProcessResult));
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.cpp:1633]: Why are you returning from here? In my last comment, I asked you to add the block to the end of the buffer in case of this error, too.

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 8]
```diff
@@ -1641,7 +1639,7 @@ OperationResult PreProcessor::launchReqPreProcessing(const PreProcessRequestMsgS
   LOG_DEBUG(
       logger(),
       "Pre-execution operation successfully completed" << KVLOG(cid, reqSeqNum, clientId, reqOffsetInBatch, blockId));
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.cpp:1641]: It is not correct to copy the block id to the (preProcessResultBuffer + request.outActualReplySize) address in case of error. Please set resultLen to request.outActualReplySize in SUCCESSFUL case and copy the block to (preProcessResultBuffer + resultLen) address.

File: tests/apollo/test_skvbc_reply.py [hunk 9]
```diff
@@ -43,7 +43,7 @@ def start_replica_cmd(builddir, replica_id):
 class SkvbcReplyTest(unittest.TestCase):
 
     @with_trio
-    @with_bft_network(start_replica_cmd)
+    @with_bft_network(start_replica_cmd, selected_configs=lambda n, f, c: c == 0 and n >= 6)
```
Comment (yuliasherman) [inline tests/apollo/test_skvbc_reply.py:46]: Change to n > 6 to run only with 7 replicas.

File: tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp [hunk 10]
```diff
@@ -147,7 +147,8 @@ void InternalCommandsHandler::preExecute(IRequestsHandler::ExecutionRequest &req
                                          concordUtils::SpanWrapper &parent_span) {
   if (req.flags & bftEngine::DB_CHECKPOINT_FLAG) return;
 
-  if (req.outExecutionStatus != static_cast<uint32_t>(OperationResult::UNKNOWN)) {
+  if (req.outExecutionStatus != static_cast<uint32_t>(OperationResult::UNKNOWN) &&
+      (req.flags & HAS_PRE_PROCESSED_FLAG)) {
```
Comment (yuliasherman) [inline tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp:151]: I don't think this additional condition is correct. We never call this function twice, like it was before with execute() function.

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 11]
```diff
@@ -1623,25 +1624,22 @@ OperationResult PreProcessor::launchReqPreProcessing(const PreProcessRequestMsgS
...
   resultLen = request.outActualReplySize;
-  if (preProcessResult != OperationResult::SUCCESS) {
-    LOG_ERROR(logger(),
-              "Pre-execution failed" << KVLOG(
-                  cid, clientId, reqOffsetInBatch, reqSeqNum, (uint32_t)preProcessResult, resultLen));
-    return preProcessResult;
-  }
-  if (request.outActualReplySize == 0) {
+
+  if (!(request.outExecutionStatus) && request.outActualReplySize == 0) {
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.cpp:1628]: 1. We need to print an error log in both cases: preProcessResult != OperationResult::SUCCESS and 
request.outActualReplySize == 0.
2. We cannot be sure that in case request.outActualReplySize is 0 the result is not SUCCESS.

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 12]
```diff
@@ -1627,21 +1628,23 @@ OperationResult PreProcessor::launchReqPreProcessing(const PreProcessRequestMsgS
     LOG_ERROR(logger(),
               "Pre-execution failed" << KVLOG(
                   cid, clientId, reqOffsetInBatch, reqSeqNum, (uint32_t)preProcessResult, resultLen));
-    return preProcessResult;
   }
-  if (request.outActualReplySize == 0) {
+
+  if (preProcessResult == OperationResult::SUCCESS && request.outActualReplySize == 0) {
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.cpp:1633]: Please leave the previous condition as is:
if (request.outActualReplySize == 0)

**Review by yuliasherman — APPROVED** [review]
LGTM!
